src/toil/job.py

```diff
@@ -1722,3 +1721,185 @@ def convertPromises(kwargs):
+                             os.path.abspath(work_dir) + ':/data']
+
+    # Ensure the user has passed a valid value for defer
+    require(defer in (None, docker_call.FORGO, docker_call.STOP, docker_call.RM),
+            'Please provide a valid value for defer.')
+
+    # Get container name which is needed for _docker_kill
+    try:
+        if any('--name=' in x for x in base_docker_call):
+            container_name = [x.split('=')[1] for x in base_docker_call if '--name=' in x][0]
+        else:
+            container_name = base_docker_call[base_docker_call.index('--name') + 1]
+    except ValueError:
+        container_name = _get_container_name(job)
+        base_docker_call.extend(['--name', container_name])
+    except IndexError:
+        raise RuntimeError("Couldn't parse Docker's `--name=` option, check parameters: " + str(base_docker_call))
+
+    # Defer the container on-exit action
+    if '--rm' in base_docker_call and defer is None:
+        defer = docker_call.RM
+    job.defer(_docker_kill, container_name, action=defer)
+    # Defer the permission fixing function.  We call this explicitly later on in this function, but
+    # we defer it as well to handle unexpected job failure.
+    job.defer(_fix_permissions, base_docker_call, tool, work_dir)
+
+    # Make subprocess call
+    call = base_docker_call + [tool] + parameters
+    job.fileStore.logToMaster("Calling docker with {}.".format(' '.join(call)))
+
+    if outfile:
+        subprocess.check_call(call, stdout=outfile)
+    else:
+        if check_output:
+            return subprocess.check_output(call)
+        else:
+            subprocess.check_call(call)
+
+    # Fix root ownership of output files
+    _fix_permissions(base_docker_call, tool, work_dir)
+
+
+def _docker_kill(container_name, action):
+    """
+    Kills the specified container.
+
+    :param str container_name: The name of the container created by docker_call
+    :param int action: What action should be taken on the container?  See `defer=` in
+           :func:`docker_call`
+    """
+    docker_call.FORGO = 0
+    docker_call.STOP = 1
+    docker_call.RM = 2
+
+    running = _container_is_running(container_name)
+    if running is None:
+        # This means that the container doesn't exist.  We will see this if the container was run
+        # with --rm and has already exited before this call.
+        logger.info('The container with name "%s" appears to have already been removed.  Nothing to '
+                    'do.', container_name)
+    else:
+        if action in (None, docker_call.FORGO):
+            logger.info('The container with name %s continues to exist as we were asked to forgo a '
+                        'post-job action on it.', container_name)
+            return
+        else:
+            logger.info('The container with name %s exists. Running user-specified defer functions.',
+                        container_name)
+            if running and action >= docker_call.STOP:
```

Comment:
If we are going to use `>=` checks on the value of action, the importance of order should be documented wherever the `docker_call` constants are defined.